### PR TITLE
Create dedicated class ResultQueryInfo for query protocol

### PR DESCRIPTION
**Pull request by radek-starburst**
## Description

When customer fetches results then `QueryInfo` object is created (if query is not finished) for every such request. `QueryInfo` object has some redundant fields that are not necessary for `query protocol`. 
Computation of these fields in the sense of cpu and memory is expensive what we observe in our benchmarks and customers (Comcast). Let's just avoid this computation. 

## Motivation
A lot of CPU cycles released

## Release notes

(x ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+@Immutable
+public class ResultQueryInfo
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java:39]: Let's make this a record

File: core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+{
+    private final QueryId queryId;
+    private final QueryState state;
+    private final boolean isScheduled;
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java:43]: Don't use `is` for boolean variable names. That's only for getters, similar to the `get` prefix.

File: core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    private final Map<String, String> setSessionProperties;
+    private final Set<String> resetSessionProperties;
+    private final Map<String, SelectedRole> setRoles;
+    private final Map<String, String> addedPreparedStatements;
+    private final Set<String> deallocatedPreparedStatements;
+    private final Optional<TransactionId> startedTransactionId;
+    private final boolean clearTransactionId;
+    private final List<TrinoWarning> warnings;
+
+    @JsonCreator
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java:64]: Why does this class need to be JSON serializable?

File: core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+public class ResultQueryInfo
+{
+    private final QueryId queryId;
+    private final QueryState state;
+    private final boolean isScheduled;
+    private final String updateType;
+    private final BasicQueryStats queryStats;
+    private final ErrorCode errorCode;
+    private final Optional<BasicStageInfo> outputStage;
+    private final boolean isFinalQueryInfo;
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java:48]: Same, don't use `is` for names

File: core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    private final ExecutionFailureInfo failureInfo;
+    private final Optional<String> setCatalog;
+    private final Optional<String> setSchema;
+    private final Optional<String> setPath;
+    private final Optional<String> setAuthorizationUser;
+    private final Boolean resetAuthorizationUser;
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java:54]: Use `boolean`

File: core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java (hunk 3 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/server/ResultQueryInfo.java:64]: It is replacement for `QueryInfo` that is also `JSON`-serializable. It is used in query protocol.

File: core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java [hunk 6]
```diff
@@ -333,9 +337,14 @@ public BasicStageStats getBasicStageStats(Supplier<Iterable<TaskInfo>> taskInfos
             physicalInputDataSize += taskStats.getPhysicalInputDataSize().toBytes();
             physicalInputPositions += taskStats.getPhysicalInputPositions();
             physicalInputReadTime += taskStats.getPhysicalInputReadTime().roundTo(NANOSECONDS);
+            physicalWrittenBytes += taskStats.getPhysicalWrittenDataSize().toBytes();
 
             internalNetworkInputDataSize += taskStats.getInternalNetworkInputDataSize().toBytes();
             internalNetworkInputPositions += taskStats.getInternalNetworkInputPositions();
+            spilledDataSize += taskStats.getPipelines().stream()
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java:344]: nit: move below `if (fragment.containsTableScanNode()) {` computation

File: core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java [hunk 7]
```diff
@@ -71,13 +74,16 @@ public class BasicStageStats
     private final int queuedDrivers;
     private final int runningDrivers;
     private final int completedDrivers;
+    private final int blockedDrivers;
     private final DataSize physicalInputDataSize;
     private final long physicalInputPositions;
+    private final DataSize physicalWrittenDataSize;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java:80]: can we move below `physicalInputReadTime`?

File: core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java [hunk 8]
```diff
@@ -100,16 +106,19 @@ public BasicStageStats(
             int queuedDrivers,
             int runningDrivers,
             int completedDrivers,
+            int blockedDrivers,
 
             DataSize physicalInputDataSize,
             long physicalInputPositions,
+            DataSize physicalWrittenDataSize,
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java:113]: move below `physicalInputReadTime`

File: core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java [hunk 9]
```diff
@@ -133,13 +142,16 @@ public BasicStageStats(
         this.queuedDrivers = queuedDrivers;
         this.runningDrivers = runningDrivers;
         this.completedDrivers = completedDrivers;
+        this.blockedDrivers = blockedDrivers;
         this.physicalInputDataSize = requireNonNull(physicalInputDataSize, "physicalInputDataSize is null");
         this.physicalInputPositions = physicalInputPositions;
+        this.physicalWrittenDataSize = requireNonNull(physicalWrittenDataSize, "physicalWrittenDataSize is null");
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java:148]: move below `physicalInputReadTime`

File: core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java [hunk 10]
```diff
@@ -214,6 +236,11 @@ public long getRawInputPositions()
         return rawInputPositions;
     }
 
+    public DataSize getSpilledDataSize()
+    {
+        return spilledDataSize;
+    }
+
     public Duration getPhysicalInputReadTime()
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java:244]: nit: this is out of place, should be below `getPhysicalInputPositions` (separate commit)

File: core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java [hunk 11]
```diff
@@ -300,13 +328,15 @@ public static BasicStageStats aggregateBasicStageStats(Iterable<BasicStageStats>
 
         long physicalInputDataSize = 0;
         long physicalInputPositions = 0;
+        long physicalWrittenBytes = 0;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java:331]: move below `physicalInputReadTime`

File: core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java [hunk 12]
```diff
@@ -338,13 +369,15 @@ public static BasicStageStats aggregateBasicStageStats(Iterable<BasicStageStats>
 
             physicalInputDataSize += stageStats.getPhysicalInputDataSize().toBytes();
             physicalInputPositions += stageStats.getPhysicalInputPositions();
+            physicalWrittenBytes += stageStats.getPhysicalWrittenDataSize().toBytes();
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java:372]: move below `physicalInputReadTime`

File: core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java [hunk 13]
```diff
@@ -365,16 +398,19 @@ public static BasicStageStats aggregateBasicStageStats(Iterable<BasicStageStats>
                 queuedDrivers,
                 runningDrivers,
                 completedDrivers,
+                blockedDrivers,
 
                 succinctBytes(physicalInputDataSize),
                 physicalInputPositions,
+                succinctBytes(physicalWrittenBytes),
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/BasicStageStats.java:405]: move below `physicalInputReadTime`

File: core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java [hunk 14]
```diff
@@ -283,13 +284,15 @@ public BasicStageStats getBasicStageStats(Supplier<Iterable<TaskInfo>> taskInfos
 
         long physicalInputDataSize = 0;
         long physicalInputPositions = 0;
+        long physicalWrittenBytes = 0;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java:287]: move below `physicalInputReadTime`

File: core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java [hunk 15]
```diff
@@ -361,16 +370,19 @@ public BasicStageStats getBasicStageStats(Supplier<Iterable<TaskInfo>> taskInfos
                 queuedDrivers,
                 runningDrivers,
                 completedDrivers,
+                blockedDrivers,
 
                 succinctBytes(physicalInputDataSize),
                 physicalInputPositions,
+                succinctBytes(physicalWrittenBytes),
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java:377]: move beloe `physicalInputReadTime`

File: core/trino-main/src/main/java/io/trino/execution/StageStats.java [hunk 16]
```diff
@@ -645,13 +646,16 @@ public BasicStageStats toBasicStageStats(StageState stageState)
                 queuedDrivers,
                 runningDrivers,
                 completedDrivers,
+                blockedDrivers,
                 physicalInputDataSize,
                 physicalInputPositions,
+                physicalWrittenDataSize,
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/StageStats.java:652]: move below `physicalInputReadTime`

**Review by sopel39 — APPROVED** [review]
lgtm % comments

**Comment by radek-starburst** [comment]
@sopel39 , done

File: core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java (hunk 14 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/StageStateMachine.java:287]: ping

**Review by sopel39 — APPROVED** [review]
small comment. Let's wait if @electrum has some comments

**Comment by sopel39** [comment]
please rebase

**Comment by sopel39** [comment]
thx!

**Pull request merged**